File: app/src/main/java/com/smartpack/packagemanager/adapters/AppOpsAdapter.java
/*
 * Copyright (C) 2021-2022 sunilpaulmathew <sunil.kde@gmail.com>
 *
 * This file is part of Package Manager, a simple, yet powerful application
 * to manage other application installed on an android device.
 *
 */

package com.smartpack.packagemanager.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.checkbox.MaterialCheckBox;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textview.MaterialTextView;
import com.smartpack.packagemanager.R;
import com.smartpack.packagemanager.utils.Common;
import com.smartpack.packagemanager.utils.RootShell;
import com.smartpack.packagemanager.utils.SerializableItems.PermissionsItems;
import com.smartpack.packagemanager.utils.ShizukuShell;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Objects;

import in.sunilpaulmathew.sCommon.CommonUtils.sCommonUtils;

/*
 * Created by Lennoard <lennoardrai@gmail.com> on Mar 14, 2021
 * Modified by sunilpaulmathew <sunil.kde@gmail.com> on Mar 17, 2021
 */
public class AppOpsAdapter extends RecyclerView.Adapter<AppOpsAdapter.ViewHolder> {

    private static ArrayList<PermissionsItems> data;

    public AppOpsAdapter(ArrayList<PermissionsItems> data) {
        AppOpsAdapter.data = data;
    }

    @NonNull
    @Override
    public AppOpsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View rowItem = LayoutInflater.from(parent.getContext()).inflate(R.layout.recycle_view_appops, parent, false);
        return new AppOpsAdapter.ViewHolder(rowItem);
    }

    @Override
    public void onBindViewHolder(@NonNull AppOpsAdapter.ViewHolder holder, int position) {
        holder.mTitle.setText(data.get(position).getTitle().toUpperCase(Locale.getDefault()));
        holder.mDescription.setText(data.get(position).getDescription());
        holder.mCheckBox.setChecked(data.get(position).isGranted());
        holder.mCheckBox.setOnClickListener(v -> {
            if (sCommonUtils.getBoolean("firstOpsAttempt", true, v.getContext())) {
                new MaterialAlertDialogBuilder(Objects.requireNonNull(v.getContext()))
                        .setIcon(R.mipmap.ic_launcher)
                        .setTitle(v.getContext().getString(R.string.warning))
                        .setMessage(v.getContext().getString(R.string.operations_warning))
                        .setCancelable(false)
                        .setPositiveButton(R.string.got_it, (dialog, id) -> {
                            sCommonUtils.saveBoolean("firstOpsAttempt", false, v.getContext());
                            holder.mCheckBox.setChecked(data.get(position).isGranted());
                        }).show();
            } else {
                if (new RootShell().rootAccess()) {
                    new RootShell().runCommand("cmd appops set " + Common.getApplicationID() + " " +
                            data.get(position).getTitle() + (data.get(position).isGranted() ? " deny" : " allow"));
                } else {
                    new ShizukuShell().runCommand("cmd appops set " + Common.getApplicationID() + " " +
                            data.get(position).getTitle() + (data.get(position).isGranted() ? " deny" : " allow"));
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final MaterialCheckBox mCheckBox;
        private final MaterialTextView mDescription, mTitle;

        public ViewHolder(View view) {
            super(view);
            this.mTitle = view.findViewById(R.id.title);
            this.mDescription = view.findViewById(R.id.description);
            this.mCheckBox = view.findViewById(R.id.checkbox);
        }
    }

}

File: app/src/main/java/com/smartpack/packagemanager/adapters/AppOpsAdapter.kt
package com.smartpack.packagemanager.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.forEach
import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.google.android.material.textview.MaterialTextView
import com.smartpack.packagemanager.databinding.ListItemOpBinding
import com.smartpack.packagemanager.utils.AppOperation

/*
 * Created by Lennoard <lennoardrai@gmail.com> on Mar 14, 2021
 */
class AppOpsAdapter(
    private val listener: OnAppOperationClickedListener
) : ListAdapter<AppOperation, BaseDataBindingViewHolder<*>>(DiffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppOpViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ListItemOpBinding.inflate(
            inflater, parent, false
        )
        return AppOpViewHolder(binding)
    }

    override fun onBindViewHolder(holder: BaseDataBindingViewHolder<*>, position: Int) {
        if (holder is AppOpViewHolder) {
            holder.bind(getItem(position))
        }
    }

    fun updateData(newList: List<AppOperation>) {
        submitList(newList)
    }

    inner class AppOpViewHolder(
        val binding: ListItemOpBinding
    ) : BaseDataBindingViewHolder<AppOperation>(binding) {
        override fun bind(item: AppOperation) {
            binding.operation = item
            binding.executePendingBindings()

            if (!item.writable) {
                binding.operationItemLayout.forEach { child ->
                    child.isEnabled = false
                }
            } else {
                binding.onAppOperationClickedListener = listener
            }
        }
    }

    companion object DiffCallback : DiffUtil.ItemCallback<AppOperation>() {
        override fun areItemsTheSame(oldItem: AppOperation, newItem: AppOperation): Boolean {
            return oldItem.name == newItem.name // Name pretty much functions as identifier here 
        }

        override fun areContentsTheSame(oldItem: AppOperation, newItem: AppOperation): Boolean {
            return oldItem == newItem
        }
    }
}

@BindingAdapter("binding:determineVisibility")
fun MaterialTextView.determineVisibility(description: String) {
    visibility = if (description.trim().isEmpty()) {
        View.GONE
    } else {
        View.VISIBLE
    }
}

interface OnAppOperationClickedListener {
    fun onAppOperationClicked(operation: AppOperation)
}